File: workers/pairfob-origin/src/room/session.ts
import { DAEMON_ID_RE, HELLO_GRACE_MS, MAX_ESTABLISHED, MAX_RESUME, PROTOCOL } from "../constants.ts";
import { isZeroRoute, newRouteId, routeHex } from "../crypto.ts";
import { encode, Typ, type Frame } from "../envelope.ts";
import { parseJSONObject, reject, sendErr, sendJSON } from "../frames.ts";
import { armHello, armResume, clearKindRef } from "./alarms.ts";
import type { RoomCore } from "./core.ts";
import type { RoomSocket } from "./types.ts";

export async function handleClientHello(room: RoomCore, ws: RoomSocket, frame: Frame): Promise<void> {
  if (!isZeroRoute(frame.routeId)) {
    reject(ws, "bad_token", "HELLO_CLIENT route_id must be zero");
    return;
  }
  const req = parseJSONObject(frame.payload);
  if (!req || req.v !== PROTOCOL || req.protocol !== PROTOCOL) {
    reject(ws, "bad_token", "invalid HELLO_CLIENT payload");
    return;
  }
  const att = room.att(ws);
  if (!att || att.role !== "phone") {
    reject(ws, "unbound", "HELLO_CLIENT on non-phone websocket");
    return;
  }
  if (att.hello_at_ms && att.mode === "hello" && att.kind === "none") {
    sendErr(ws, "unbound", "duplicate HELLO_CLIENT");
    ws.close(1000, "unbound");
    return;
  }
  const now = room.now();
  att.hello_at_ms = now;
  att.mode = "hello";
  room.writeAtt(ws, att);
  await armHello(room, String(now), now);
}

export async function handleSessionAttach(room: RoomCore, ws: RoomSocket, frame: Frame): Promise<void> {
  if (!isZeroRoute(frame.routeId)) {
    reject(ws, "bad_token", "invalid SESSION_ATTACH payload");
    return;
  }
  const req = parseJSONObject(frame.payload);
  const daemonId = typeof req?.daemon_id === "string" ? req.daemon_id : "";
  if (!req || req.v !== PROTOCOL || !DAEMON_ID_RE.test(daemonId)) {
    reject(ws, "bad_token", "invalid SESSION_ATTACH payload");
    return;
  }
  const att = room.att(ws);
  if (!att || att.role !== "phone") {
    reject(ws, "unbound", "SESSION_ATTACH on non-phone websocket");
    return;
  }
  if (!att.hello_at_ms || room.now() - att.hello_at_ms > HELLO_GRACE_MS) {
    sendErr(ws, "unbound", "SESSION_ATTACH after HELLO timeout");
    ws.close(1000, "unbound");
    return;
  }
  if (att.mode === "pairing" || att.kind === "pairing") {
    sendErr(ws, "wrong_ws", "SESSION_ATTACH on PairingWS");
    return;
  }
  if (att.kind === "resumehello" || att.kind === "established" || att.mode === "session") {
    sendErr(ws, "wrong_ws", "SessionWS already attached");
    return;
  }
  if (daemonId !== room.daemonId) {
    reject(ws, "bad_frame", "SESSION_ATTACH daemon_id mismatch");
    return;
  }
  if (!room.daemon) {
    sendErr(ws, "daemon_offline", "no daemon");
    return;
  }

  const { resume } = room.countKinds();
  if (resume >= MAX_RESUME) {
    const lru = room.lruResume();
    if (lru) room.closeBind(lru, "kicked", "replaced LRU ResumeHello");
  }

  const rid = newRouteId((n) => room.random(n));
  const hex = routeHex(rid);
  const now = room.now();
  att.mode = "session";
  att.kind = "resumehello";
  att.route_id = hex;
  att.created_ms = now;
  room.writeAtt(ws, att);
  room.store.upsertBind({ route_id: hex, kind: "resumehello", created_at: now, pair_ref: "" });
  room.noteBind("resumehello");
  await armResume(room, hex, now);

  const body = { v: PROTOCOL, route_id: hex };
  sendJSON(ws, Typ.SESSION_BOUND, rid, body);
  sendJSON(room.daemon, Typ.SESSION_BOUND, rid, body);
}

export async function handleSessionEstablished(room: RoomCore, ws: RoomSocket, frame: Frame): Promise<void> {
  if (room.daemon !== ws) return;
  const req = parseJSONObject(frame.payload);
  const want = routeHex(frame.routeId);
  const got = typeof req?.route_id === "string" ? req.route_id : "";
  if (isZeroRoute(frame.routeId) || !req || req.v !== PROTOCOL || got !== want) {
    sendErr(ws, "unbound", "SESSION_ESTABLISHED route mismatch", { routeId: frame.routeId });
    return;
  }
  const client = room.findByRoute(want);
  const att = client ? room.att(client) : null;
  if (!client || !att || att.kind !== "resumehello") {
    sendErr(ws, "unbound", "route is not ResumeHello", { routeId: frame.routeId });
    return;
  }
  const { est } = room.countKinds();
  if (est >= MAX_ESTABLISHED) {
    room.closeBind(client, "too_many_devices", "established cap 10");
    return;
  }
  att.kind = "established";
  room.writeAtt(client, att);
  room.store.upsertBind({ route_id: att.route_id, kind: "established", created_at: att.created_ms, pair_ref: "" });
  room.noteBind("established");
  await clearKindRef(room.store, "resume_15s", att.route_id);
  client.send(encode(frame));
}

export function handleFakeEstablished(ws: RoomSocket): void {
  sendErr(ws, "unbound", "clients must not send SESSION_ESTABLISHED");
}

export function handleDaemonError(room: RoomCore, ws: RoomSocket, frame: Frame): void {
  const req = parseJSONObject(frame.payload);
  const code = typeof req?.code === "string" ? req.code : "";
  if (!req || !code) {
    sendErr(ws, "bad_token", "invalid ERROR payload");
    return;
  }
  if (room.daemon !== ws) return;
  const bodyRid = typeof req.route_id === "string" ? req.route_id : "";
  if (bodyRid) {
    const want = routeHex(frame.routeId);
    if (bodyRid !== want) {
      sendErr(ws, "bad_token", "ERROR route mismatch");
      return;
    }
    const client = room.findByRoute(want);
    if (!client) return;
    client.send(encode(frame));
    const att = room.att(client);
    if (att?.route_id) room.deleteBind(att.route_id);
    try {
      client.close(1000, code);
    } catch {
      /* already closed */
    }
    return;
  }
  if (!isZeroRoute(frame.routeId)) {
    sendErr(ws, "bad_token", "ERROR missing route_id payload");
    return;
  }
  for (const c of room.sockets()) {
    const a = room.att(c);
    if (a?.role === "phone") c.send(encode(frame));
  }
}

File: internal/mux/session.go
package mux

import (
	"encoding/hex"
	"time"

	"pairfob/internal/envelope"
)

func (h *Hub) clientHello(st *clientState, f envelope.Frame) {
	var req struct {
		V        int `json:"v"`
		Protocol int `json:"protocol"`
	}
	if !isZeroRID(f.RouteID) || decodeJSON(f.Payload, &req) != nil || req.V != 1 || req.Protocol != 1 {
		h.reject(st.conn, "bad_token", "invalid HELLO_CLIENT payload")
		return
	}
	h.mu.Lock()
	defer h.mu.Unlock()
	if st.hello {
		h.sendErr(st.conn, envelope.ErrorBody{Code: "unbound", Message: "duplicate HELLO_CLIENT"})
		st.conn.Close()
		return
	}
	st.hello = true
	st.helloAt = h.now()
	st.timer = time.AfterFunc(h.helloGrace, func() {
		h.mu.Lock()
		defer h.mu.Unlock()
		if current := h.clients[st.conn]; current == st && st.mode == "" {
			h.sendErr(st.conn, envelope.ErrorBody{Code: "unbound", Message: "5s attach timeout"})
			st.conn.Close()
			delete(h.clients, st.conn)
		}
	})
}

func (h *Hub) sessionAttach(st *clientState, f envelope.Frame) {
	var req struct {
		V        int    `json:"v"`
		DaemonID string `json:"daemon_id"`
	}
	if !isZeroRID(f.RouteID) || decodeJSON(f.Payload, &req) != nil || req.V != 1 || !daemonIDPattern.MatchString(req.DaemonID) {
		h.reject(st.conn, "bad_token", "invalid SESSION_ATTACH payload")
		return
	}
	h.mu.Lock()
	if !st.hello || h.now().Sub(st.helloAt) > h.helloGrace {
		h.mu.Unlock()
		h.sendErr(st.conn, envelope.ErrorBody{Code: "unbound", Message: "SESSION_ATTACH after HELLO timeout"})
		st.conn.Close()
		return
	}
	if st.mode == "pairing" {
		h.mu.Unlock()
		h.sendErr(st.conn, envelope.ErrorBody{Code: "wrong_ws", Message: "SESSION_ATTACH on PairingWS"})
		return
	}
	if st.bind != nil || st.mode == "session" {
		h.mu.Unlock()
		h.sendErr(st.conn, envelope.ErrorBody{Code: "wrong_ws", Message: "SessionWS already attached"})
		return
	}
	d := h.daemons[req.DaemonID]
	if d == nil || d.conn == nil {
		h.mu.Unlock()
		h.sendErr(st.conn, envelope.ErrorBody{Code: "daemon_offline", Message: "no daemon"})
		return
	}
	var pending []pendingSend
	_, nResume := countKinds(d)
	if nResume >= 2 {
		var lru *bind
		for _, b := range d.binds {
			if b.kind == kindResumeHello && (lru == nil || b.created.Before(lru.created)) {
				lru = b
			}
		}
		if lru != nil {
			pending = append(pending, h.kickBindLocked(lru, "kicked", "replaced LRU ResumeHello")...)
		}
	}
	rid, err := randRID()
	if err != nil {
		h.mu.Unlock()
		flushPending(pending)
		h.sendErr(st.conn, envelope.ErrorBody{Code: "forbidden", Message: "relay entropy unavailable"})
		return
	}
	b := &bind{id: rid, client: st.conn, daemon: d, kind: kindResumeHello, created: h.now()}
	b.timer = time.AfterFunc(h.resumeWait, func() {
		h.mu.Lock()
		var late []pendingSend
		if b.kind == kindResumeHello && b.daemon.binds[b.id] == b {
			late = h.queueCloseBindLocked(b, routeErrFrame(b, "unpaired", "15s DeviceHello timeout"), true)
		}
		h.mu.Unlock()
		flushPending(late)
	})
	d.binds[rid] = b
	st.bind = b
	st.daemon = d
	st.mode = "session"
	if st.timer != nil {
		st.timer.Stop()
		st.timer = nil
	}
	client, daemonConn := st.conn, d.conn
	body := map[string]any{"v": 1, "route_id": hex.EncodeToString(rid[:])}
	bound := envelope.JSON(envelope.TypSESSION_BOUND, rid, body)
	h.mu.Unlock()
	flushPending(pending)
	if err := client.Send(bound); err != nil {
		h.mu.Lock()
		var fail []pendingSend
		if d.binds[b.id] == b {
			fail = h.queueCloseBindLocked(b, routeErrFrame(b, "unpaired", "client send failed"), true)
		}
		h.mu.Unlock()
		flushPending(fail)
		return
	}
	if err := daemonConn.Send(bound); err != nil {
		h.mu.Lock()
		var fail []pendingSend
		if d.conn == daemonConn {
			fail = h.dropDaemonLocked(d)
		}
		h.mu.Unlock()
		flushPending(fail)
	}
}

func (h *Hub) sessionEstablished(c Conn, f envelope.Frame) {
	var req struct {
		V       int    `json:"v"`
		RouteID string `json:"route_id"`
	}
	wantRID := hex.EncodeToString(f.RouteID[:])
	if isZeroRID(f.RouteID) || decodeJSON(f.Payload, &req) != nil || req.V != 1 || req.RouteID != wantRID {
		h.sendRouteErr(c, f.RouteID, envelope.ErrorBody{Code: "unbound", RouteID: wantRID, Message: "SESSION_ESTABLISHED route mismatch"})
		return
	}
	h.mu.Lock()
	d := h.daemonByConn(c)
	if d == nil {
		h.mu.Unlock()
		return
	}
	b := d.binds[f.RouteID]
	if b == nil || b.kind != kindResumeHello {
		h.mu.Unlock()
		h.sendRouteErr(c, f.RouteID, envelope.ErrorBody{Code: "unbound", RouteID: wantRID, Message: "route is not ResumeHello"})
		return
	}
	nEst, _ := countKinds(d)
	if nEst >= 10 {
		pending := h.queueCloseBindLocked(b, routeErrFrame(b, "too_many_devices", "established cap 10"), true)
		h.mu.Unlock()
		flushPending(pending)
		return
	}
	if b.timer != nil {
		b.timer.Stop()
		b.timer = nil
	}
	b.kind = kindEstablished
	client := b.client
	h.mu.Unlock()
	if err := client.Send(f); err != nil {
		h.mu.Lock()
		var pending []pendingSend
		if d.binds[b.id] == b {
			pending = h.queueCloseBindLocked(b, routeErrFrame(b, "unpaired", "client send failed"), true)
		}
		h.mu.Unlock()
		flushPending(pending)
	}
}

func (h *Hub) errorFromDaemon(c Conn, f envelope.Frame) {
	var body envelope.ErrorBody
	if decodeJSON(f.Payload, &body) != nil || body.Code == "" {
		h.sendErr(c, envelope.ErrorBody{Code: "bad_token", Message: "invalid ERROR payload"})
		return
	}
	h.mu.Lock()
	d := h.daemonByConn(c)
	if d == nil {
		h.mu.Unlock()
		return
	}
	var pending []pendingSend
	if body.RouteID != "" {
		if !isLowerHex(body.RouteID, 16) || body.RouteID != hex.EncodeToString(f.RouteID[:]) {
			h.mu.Unlock()
			h.sendErr(c, envelope.ErrorBody{Code: "bad_token", Message: "ERROR route mismatch"})
			return
		}
		if b := d.binds[f.RouteID]; b != nil {
			pending = h.queueCloseBindLocked(b, f, false)
		}
		h.mu.Unlock()
		flushPending(pending)
		return
	}
	if !isZeroRID(f.RouteID) {
		h.mu.Unlock()
		h.sendErr(c, envelope.ErrorBody{Code: "bad_token", Message: "ERROR missing route_id payload"})
		return
	}
	for _, b := range d.binds {
		pending = append(pending, pendingSend{c: b.client, f: f})
	}
	h.mu.Unlock()
	flushPending(pending)
}

func countKinds(d *daemonSlot) (est, resume int) {
	for _, b := range d.binds {
		switch b.kind {
		case kindEstablished:
			est++
		case kindResumeHello:
			resume++
		}
	}
	return
}

// BindKindCounts is the live mux table (tests assert ResumeHello cap).
func (h *Hub) BindKindCounts(daemonID string) (established, resumeHello int) {
	h.mu.Lock()
	defer h.mu.Unlock()
	d := h.daemons[daemonID]
	if d == nil {
		return 0, 0
	}
	return countKinds(d)
}
